freqtrade/optimize/backtesting.py

```diff
@@ -344,6 +344,8 @@ def backtest(self, processed: Dict, stake_amount: float,
                         and tmp != end_date
                         and row[BUY_IDX] == 1 and row[SELL_IDX] != 1):
                     # Enter trade
+                    prev_date = data[pair][indexes[pair] - 1][DATE_IDX]
+                    stake_amount = get_stake_amount(pair, prev_date)
```

Comment:
What's the reasoning behind this? (i don't think it's correct, that's why i'm asking) 

Candle dates are from the candle-start - and we use the candle's open-price as entry-price - so assuming a 1h candle, you'd now use the date of 10:00 for a buy-signal that would use the open-signal from 11:00 ...

Reply:
signals are shifted

Reply:
signals are shifted - yes - but shifting price back is wrong as that will "nullify" the shift partially.

Consider a current 1h candle - it's now 15:22 (UTC) - so the hourly candle will have a timestamp of 14:00.
The signal will however only be available since 15:00:01 - as this is when the candle "closed" and the signal was generated.

so a "real" trade will be bought at 15:00:01 - based on data from the candle 14:00-15:00 - ideally with a price close to the candle's open price (the candle from 15:00 - 16:00) - (hence we're using "open" as price for backtesting) 

This is the reason the signal must be shifted for backtesting - as the "date" always refers to the candle start - not the candle end - so a signal based on data for the candle at 14:00 will never be acted before the candle is complete.

---

Now thinking about this in the backtesting context: 

The signal based on 14:00-15:00 is acted on at 15:00:01 - so by shifting the date back, we call `get_stake_amount()` with 14:00 - while the open-price is taken from 15:00-16:00 - and in real mode, 15:00 is the earliest that signal did appear (the candle is incomplete up to that point, and therefore not available to the strategy to avoid repainting).

Now depending on what you do with this date, it may not be relevant - but to have this date close to how it would be called in reality, it should be called with 15:00 - the earliest time a "live" signal would be acted on

Reply:
This boils down on what's more intuitive, I think it's simpler to assume that signals and stake amount are aligned, it doesn't matter if dates are shifted, from a backtesting point of view, you have to ask for the index where that signal was populated. 
During live it would indeed be confusing if the call would ask for the previous date, but the date is not given in live mode...so 🤷 

If you name the variable like `candle_date` then I guess its more clear that it is supposed to be the date of the candle, and NOT the date when the callback _would_ be executed

Reply:
let's leave it for now - and if we're seeing problems / confusion, this can be adjusted eventually...